buildpacks/nodejs-yarn/src/main.rs

```diff
@@ -0,0 +1,156 @@
+#![warn(unused_crate_dependencies)]
+#![warn(clippy::pedantic)]
+#![warn(clippy::cargo)]
+#![allow(clippy::module_name_repetitions)]
+
+use crate::layers::{DistLayer, DistLayerError};
+use heroku_nodejs_utils::inv::Inventory;
+use heroku_nodejs_utils::package_json::{PackageJson, PackageJsonError};
+use heroku_nodejs_utils::vrs::Requirement;
+use libcnb::build::{BuildContext, BuildResult, BuildResultBuilder};
+use libcnb::data::build_plan::BuildPlanBuilder;
+use libcnb::data::launch::{Launch, ProcessBuilder};
+use libcnb::data::{layer_name, process_type};
+use libcnb::detect::{DetectContext, DetectResult, DetectResultBuilder};
+use libcnb::generic::GenericMetadata;
+use libcnb::generic::GenericPlatform;
+use libcnb::{buildpack_main, Buildpack};
+use libherokubuildpack::{log_error, log_header, log_info};
+use thiserror::Error;
+
+mod layers;
+
+const INVENTORY: &str = include_str!("../inventory.toml");
+
+pub struct NodeJsYarnBuildpack;
+
+impl Buildpack for NodeJsYarnBuildpack {
+    type Platform = GenericPlatform;
+    type Metadata = GenericMetadata;
+    type Error = NodeJsYarnBuildpackError;
+
+    fn detect(&self, context: DetectContext<Self>) -> libcnb::Result<DetectResult, Self::Error> {
+        context
+            .app_dir
+            .join("yarn.lock")
+            .exists()
+            .then(|| {
+                DetectResultBuilder::pass()
+                    .build_plan(
+                        BuildPlanBuilder::new()
+                            .provides("yarn")
+                            .requires("yarn")
+                            .provides("node_modules")
+                            .requires("node_modules")
```

Comment:
2 questions here:
* what does the `node_modules` buildpack do?  
* would the PnP feature from Yarn 2 require this?

Reply:
These aren't buildpacks per se. This is basically a list of things that this buildpack provides and requires. These strings are used to evaluate whether a group of buildpacks are compatible and should pass detection. 

This buildpack requires `node`, so it won't pass detection unless another buildpack in the group provides `node` (which heroku/nodejs-engine does). 

This buildpack provides `node_modules`, which would allow any other buildpack to require `node_modules`, which is a way to say "My buildpack expects another buildpack to install node_modules". It's also requiring `node_modules` so that it will pass detection, even if no other buildpack in the group is asking for `node_modules`.

I mostly chose to go with `node_modules` because 1) that's what the bash buildpack was doing 2) the paketo buildpacks also use `node_modules` and 3) I kind of thought the distinction between a pnp.js and a `node_modules` folder weren't relevant for downstream buildpacks -- they both provide modules for node to use.